automaton.py

```diff
@@ -1,37 +1,140 @@
+    def __init__(self, dest: str, target: list=[], symbol: list=[]):
         self.dest = dest
         self.target = target
         self.symbol = symbol
 
     def render(self):
         # TODO: Return a string in the format s0:a>s1
         pass
+        
+class operationID(enum.Enum):
+    INVALID = 0
+    STATES = 1
+    INITIAL = 2
+    ACCEPT = 3
+    ALPHABET = 4
+    TRANSITIONS = 5
+    
+    
 
 class Automaton():
 
     def __init__(self, states: list = [], initial: str = "", accept: list = [], alphabet: list = [], transitions: list=[], from_file: str = ""):
-        self.states = states
-        self.initial = initial
-        self.accept = accept
-        self.alphabet = alphabet
-        self.transitions = transitions
+        self.states = []
+        self.initial = ""
+        self.accept = []
+        #include "$" (epsilon transition) in alphabet
+        self.alphabet = ['$']
+        self.transitions = []
         self.is_valid = False
         self.errors = []
         if from_file:
+            self.initial = ""
             self._parse_file(from_file)
         self.validate()
 
     def validate(self):
         # TODO: Validate if automaton is valid, if so, set self.is_valid = True
         # if not, add errors to self.errors
-        pass
+        
+        #"set" help to see if an element is repeated (if lenght is different from the lenght of the set, this means a element is repeated)
+        #states repeated, throw an error
+        if len(self.states) != len(set(self.states)):
+            raise NotImplementedError
+        #initial state is undefined
+        if (self.initial == "" or self.initial not in self.states):
+            raise NotImplementedError
+        #accept state has to exist in states
+        for sa in self.accept:
+            if(sa not in self.states):
+                raise NotImplementedError
+        #alphabet cannot be repeated
+        if len(self.alphabet) != len(set(self.alphabet)):
+            raise NotImplementedError 
+        #validate if states of transitions are valid
+        for trans in self.transitions:
+            if (trans.dest not in self.states):
+                print("Invalid destination state")
+                raise NotImplementedError
+            for tar in trans.target:
+                if (tar not in self.states):
+                    print("Invalid target state")
+                    raise NotImplementedError
+            for alpha in trans.symbol:
+                if (alpha not in self.alphabet):
+                    print("Undefined symbol in alphabet: ", alpha)
+                    raise NotImplementedError
+                    
 
     def render(self):
         # TODO: Return a string in a format accepted by FSM simulator
         pass
 
     def _parse_file(self, filename):
         # TODO: Fill attributes accordingly to what was read from file
-        pass
+            parsingID = operationID.INVALID
+            f = open(filename,"r")
```

Comment:
use the keyword `with` to avoid keeping the fd open when it's not necessary:

```python
with open(filename, "r") as f:
    for x in f:
        ...
```